src/plans/constrained_plan.jl

```diff
@@ -1,28 +1,42 @@
 @doc raw"""
-    ConstraintType
+    AbstractConstraintType
 
 An abstract type to represent different forms of representing constraints
+as well as their gradients.
 """
-abstract type ConstraintType end
+abstract type AbstractConstraintType end
 
 @doc raw"""
-    FunctionConstraint <: ConstraintType
+    FunctionConstraint{CT <: AbstractConstraintType}  <: AbstractConstraintType
 
 A type to indicate that constraints are implemented one whole functions,
 for example ``g(p) ∈ ℝ^m``.
+
+For the gradient there are two possible variants available:
+
+* [`PowerManifoldVectorialType`](@ref): ``\operatorname{grad} g\colon \mathcal M \to (T_p\mathcal M)^m``,
+  the gradient returns a vector of gradients, one for each component function of ``g``.
+* `PowerManifoldTangentConstaint`: ``\operatorname{grad} g\colon \mathcal M \to T_P\mathcal M^m``,
+  where ``P = (p,…,p) \in\mathcal M^m``, that is
+  the gradient returns a tangent vector on the power manifold.
 """
-struct FunctionConstraint <: ConstraintType end
+struct FunctionConstraint{CT<:AbstractVectorialType} <: AbstractConstraintType
+    constraint_type::CT
+end
+
+function FunctionConstraint(power_representation::AbstractPowerRepresentation=NestedPowerRepresentation())
```

Comment:
HM, this would maybe be breaking? Ah with the nested case as a default maybe not, since that was the default before.

Reply:
I think it shouldn't be breaking with nested as default.

Reply:
Probably not, you are right. We just have to get this internally correct then now ;) But for the next 2-3 days I might be a bit busy correcting my exam. Feel free to check the ones I implemented already, I will continue with single gradient access and such later then.

Reply:
No problem. I originally thought I could jump straight to issues with power manifold but Manopt didn't even compile and I still have failures well before that part so I will probably wait with more changes unless you have a piece of code that I could run to test PowerManifold-related improvements.

Reply:
Ys sure it did not compile, I am currently not able to integrate the idea of a (non-nested) power manifold at all. I thought I had an idea but none of that works. So then I could not get it to work and hence asked for help. I currently have not a single idea how to get that to work and sure the current state of this PR is a bit of a chaos (since basically nothing works in the new representation idea).
And sure when nothing works, there is also no way to test any improvements ;) That trail of thoughts I can easily follow.

Reply:
I see. I don't think I can help much with the vector-of-functions design because it doesn't mesh well with the non-nested power manifold approach.

Reply:
Hm, but then this would be super-breaking if we loose that one.

And as explained below – single (component) access of the constraints might be super-crucial in having ALM/EPM fast, so I really do not want to “loose” those. For now I do not have a good idea how to have both, then.

Reply:
> And as explained below – single (component) access of the constraints might be super-crucial in having ALM/EPM fast, so I really do not want to “loose” those. For now I do not have a good idea how to have both, then.

I have presented you the exact idea how to get both...

Reply:
Nothing is lost, you'd just write `co.g(M, p, i)` instead of `co.g[i](M, p)`.